web_sdk/BUILD.gn

```diff
@@ -43,60 +32,100 @@ group("web_sdk") {
+
+  source_dart_files = exec_script("//third_party/dart/tools/list_dart_files.py",
+                                  [
+                                    "absolute",
+                                    rebase_path(invoker.input_dir),
+                                  ],
+                                  "list lines")
+
+  if (defined(invoker.exclude_patterns)) {
+    filtered_files = filter_exclude(source_dart_files, invoker.exclude_patterns)
+    source_dart_files = []
+    source_dart_files = filtered_files
+  }
 
-  packages = dart_sdk_package_config
-  script = "sdk_rewriter.dart"
-  output_dir = rebase_path("$root_out_dir/flutter_web_sdk/lib/ui/")
-  input_dir = rebase_path("//flutter/lib/web_ui/lib/")
-  pool = "//flutter/build/dart:dart_pool"
+  prebuilt_dart_action(target_name) {
+    packages = dart_sdk_package_config
+    pool = "//flutter/build/dart:dart_pool"
 
-  outputs = [
-    "$target_gen_dir/$target_name.stamp",
-    "$root_out_dir/flutter_web_sdk/lib/ui/",
-  ]
+    script = "sdk_rewriter.dart"
+    inputs = source_dart_files + [ "sdk_rewriter.dart" ]
 
-  args = [
-    "--output-dir=$output_dir",
-    "--input-dir=$input_dir",
-    "--ui",
-    "--stamp",
-    rebase_path(outputs[0], root_build_dir),
-  ]
+    stamp_location = "$target_gen_dir/$target_name.stamp"
+    outputs = [
+      stamp_location,
+      invoker.output_dir,
+    ]
+
+    input_dir = rebase_path(invoker.input_dir)
+    output_dir = rebase_path(invoker.output_dir)
+
+    args = [
+      "--output-dir=$output_dir",
+      "--input-dir=$input_dir",
+      "--stamp",
+      rebase_path(stamp_location, root_build_dir),
+    ]
+    if (ui) {
+      args += [ "--ui" ]
+    } else {
+      library_name = invoker.library_name
+      api_file = rebase_path(invoker.api_file)
+      args += [
+        "--library-name=$library_name",
+        "--api-file=$api_file",
+      ]
+    }
 
-  foreach(source_file, web_ui_sources) {
-    path = rebase_path(source_file)
-    args += [ "--input=$path" ]
+    foreach(source_file, source_dart_files) {
+      path = rebase_path(source_file)
+      args += [ "--source-file=$path" ]
+    }
   }
 }
 
-prebuilt_dart_action("web_engine_sources") {
-  inputs = web_engine_sources + [ "sdk_rewriter.dart" ]
+sdk_rewriter("web_ui_library_sources") {
+  ui = true
+  input_dir = "//flutter/lib/web_ui/lib/"
+  output_dir = "$root_out_dir/flutter_web_sdk/lib/ui/"
 
-  packages = dart_sdk_package_config
-  script = "sdk_rewriter.dart"
-  output_dir = rebase_path("$root_out_dir/flutter_web_sdk/lib/_engine/")
-  input_dir = rebase_path("//flutter/lib/web_ui/lib/src/")
-  pool = "//flutter/build/dart:dart_pool"
+  # exclude everything in the engine directory, it will be a separate internal library
+  exclude_patterns = [ rebase_path("//flutter/lib/web_ui/lib/src/*") ]
+}
 
-  outputs = [
-    "$target_gen_dir/$target_name.stamp",
-    "$root_out_dir/flutter_web_sdk/lib/_engine/",
-  ]
+sdk_rewriter("web_engine_library") {
```

Comment:
nit: for consistency with "ui", should this also end with `_sources`?

Reply:
Actually no! I did this purposely because I find out quite confusing before, because there was both a variable called `web_ui_sources` (the list of paths) and a target called `web_ui_sources` (which was the action to call sdk_rewriter). Now, I've changed it so that the targets end in `library` and the variable that contains the actual sources ends in `sources`.